crates/oxc_transformer/src/es2015/arrow_functions.rs

```diff
@@ -137,22 +137,35 @@ impl<'a> Traverse<'a> for ArrowFunctions<'a> {
             return;
         }
 
-        let ident = match name {
-            JSXElementName::Identifier(ident) => ident,
-            JSXElementName::MemberExpression(member_expr) => {
-                member_expr.get_object_identifier_mut()
+        if let JSXElementName::Identifier(ident) = name {
+            if ident.name == "this" {
+                // We can't produce a proper identifier with a `ReferenceId` because `JSXIdentifier`
+                // lacks that field. https://github.com/oxc-project/oxc/issues/5352
+                // So generate a reference and just use its name.
+                // If JSX transform is enabled, that transform runs before this and will have converted
+                // this to a proper `ThisExpression`, and this visitor won't run.
+                // So only a problem if JSX transform is disabled.
+                let new_ident = self.get_this_name(ctx).create_read_reference(ctx);
+                ident.name = new_ident.name;
```

Comment:
We *can* now produce a proper identifier!

```suggestion
                let mut new_ident = self.get_this_name(ctx).create_read_reference(ctx);
                new_ident.span = ident.span;
                *name = JSXElementName::IdentifierReference(ctx.ast.alloc(new_ident));
```